_rules/heading-name-is-not-only-whitespace-ffd0e9.md

```diff
@@ -0,0 +1,185 @@
+
+## Assumptions
+
+This rule assumes that having an element that unintentionally shows up programmatically to a user of assistive technologies as a heading element, but is not shown visually as a heading on the page, is a violation of WCAG Success Criterion 1.3.1 Info and Relationships.
+
+## Accessibility Support
+
+Handling of headings containing only whitespace characters, carriage returns, newlines, tabs, and form-feeds varies between different assistive technologies and browsers. This means that even though the outcome of this rule is *failed*, users of certain assistive technology and browser combinations might not experience an issue.
+
+## Background
+
+- In some screen reader and browser combinations, headings containing only whitespace characters, carriage returns, newlines, tabs, and form-feeds will show up as empty headings, confusing the user experience. For a screen reader user it will be hard to get an overview of the heading structure of the page if "empty" headings are included in the heading structure. Indeed, it can be hard to know if the heading is just whitespace that has accidentally been marked up as a heading, or if it is an actual heading that for some reason doesn't have an accessible name. Since this is a case where the programmatically determinable structure of the page doesn't match the visual presentation, this is a violation of success criterion [1.3.1 Info and Relationships](https://www.w3.org/TR/WCAG21/#info-and-relationships).
+- [Understanding Success Criterion 1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
+
+## Test Cases
+
+### Passed
+
+#### Passed Example 1
+
+`h2` element with no [accessible name](#accessible-name).
+
+```html
+<h2></h2>
+```
+
+#### Passed Example 2
+
+Element with the semantic role of heading has no [accessible name](#accessible-name).
+
+```html
+<div role="heading"></div>
+```
+
+#### Passed Example 3
+
+`h2` element has an [accessible name](#accessible-name) from content that is not only whitespace.
+
+```html
+<h2>' </h2>
+```
+
+#### Passed Example 4
+
+`h2` element has an [accessible name](#accessible-name) through `aria-label` that is not only whitespace.
+
+```html
+<h2 aria-label="Orange harvesting season"> </h2>
+```
+
+#### Passed Example 5
+
+`h2` element has an [accessible name](#accessible-name) through the `alt` attribute that is not only whitespace.
+
+```html
+<h2><img src="#" alt="Orange harvesting season"> </h2>
+```
+
+#### Passed Example 6
+
+`h2` element contains the phrasing content element `<span>` as only content, and this does not affect the accessible name computation.
+
+```html
+<h2><span></span></h2>
+```
+
+### Failed
+
+#### Failed Example 1
+
+`h2` element contains `br` element as only content. This is translated into a single space in the [accessible name](#accessible-name) computation, which gives an [accessible name](#accessible-name) that is only whitespace.
+
+```html
+<h2><br /></h2>
```

Comment:
This example doesn't fail the rule. The accessible name of this heading is an empty string (""), exactly the same as passed example 1. The accessible name is always trimmed space characters before its returned. For example, take a look at this:

```html
<element1 id="el1" aria-labelledby="el3" />
<element2 id="el2" aria-labelledby="el1" />
<element3 id="el3"> hello </element3>
```

From: https://www.w3.org/TR/accname-1.1/#ex-1-example-1-element1-id-el1-aria-labelledby-el3-element2-id-el2-aria-labelledby-el1-element3-id-el3-hello-element3
> element1's accessible name is "hello" because this is a first traversal of its aria-labelledby, leading to element3.

This isn't well documented by any means, but every implementation of the accessible name computation algorithm  I've researched does this. This example also clearly shows it. There are also others in the accName test cases https://www.w3.org/wiki/AccName_1.1_Testable_Statements

Reply:
I do not see which part of the algorithm is supposed to do that trimming 😞 
Especially, steps 2F.iii and 2F.iv (name from content) and 2G (text node) are just appending text, not trimming it of whitespace.

At the same time, I do see this happening in tests I do… 😕 

---
Also, the algorithm is supposed to return a flat string (newlines and such converted to spaces, adjacent spaces reduced to a single space) as per the first paragraph of https://www.w3.org/TR/accname-1.1/#mapping_additional_nd
> The output is a flat, unstructured string

but I do not see which part of the algorithm is flattening the output (except for step 2D which handle `title` and similar attributes. Notably, step 2G (text node) does not flatten the string, which is weird.

---
In any case, if accessible name is always trimmed of spaces, then it can never be only whitespace and this rule is pointless. Many other rules also have weird formulation of "having an accessible name which is not only whitespace" and need to be updated. I'll make an issue on that…

---
Running some tests, with headers whose content is various empty whitespace (`<h1>something</h1>`), is "fun", as always…
* When `something` is (nothing), ` ` (1 space), `  ` (2 spaces), (newline), `<br />`: both Chrome, Firefox and Edge give it an empty accessible name (actually, Edge gives no accessible name and Firefox gives an accessible name of `null` 🙄 ).
* When `something` is `&nbsp;`, Firefox and Edge handle it as the other cases, but Chrome gives it an accessible name of `" "` (which I'm not sure if it's a non breaking space or a normal space, my quick test shows a normal space but that might just be my copy/paste changing it…)
* When `something` is `&nbsp;` or `<br />`, both Chrome, Firefox and Edge render a large vertical space (the space needed for `h1`) but for the other cases, none of them render any extra space. Not really related to the current rule or the accessible name, but still "fun"…

Reply:
> In any case, if accessible name is always trimmed of spaces, then it can never be only whitespace and this rule is pointless. Many other rules also have weird formulation of "having an accessible name which is not only whitespace" and need to be updated. I'll make an issue on that…

See #976